source/steamshim/src/os.cpp

```diff
@@ -0,0 +1,214 @@
+#include "os.h"
+#include "parent/parent_private.h"
+#include <cstring>
+
+#ifdef _WIN32
+void fail(const char *err)
+{
+    MessageBoxA(NULL, err, "ERROR", MB_ICONERROR | MB_OK);
+    ExitProcess(1);
+} // fail
+
+bool createPipes(PipeType *pPipeParentRead, PipeType *pPipeParentWrite,
+                        PipeType *pPipeChildRead, PipeType *pPipeChildWrite)
+{
+    SECURITY_ATTRIBUTES pipeAttr;
+
+    pipeAttr.nLength = sizeof (pipeAttr);
+    pipeAttr.lpSecurityDescriptor = NULL;
+    pipeAttr.bInheritHandle = TRUE;
+    if (!CreatePipe(pPipeParentRead, pPipeChildWrite, &pipeAttr, 0))
+        return 0;
+
+    pipeAttr.nLength = sizeof (pipeAttr);
+    pipeAttr.lpSecurityDescriptor = NULL;
+    pipeAttr.bInheritHandle = TRUE;
+    if (!CreatePipe(pPipeChildRead, pPipeParentWrite, &pipeAttr, 0))
+    {
+        CloseHandle(*pPipeParentRead);
+        CloseHandle(*pPipeChildWrite);
+        return 0;
+    } // if
+
+    return 1;
+} // createPipes
+
+bool setEnvVar(const char *key, const char *val)
+{
+    return (SetEnvironmentVariableA(key, val) != 0);
+} // setEnvVar
+
+bool launchChild(ProcessType *pid)
+{
+    LPWSTR str = _wcsdup( GetCommandLineW() );
+    STARTUPINFOW si = { sizeof( si ) };
+
+    memset( pid, 0, sizeof( *pid ) );
+    bool bResult = ( CreateProcessW( TEXT( ".\\" ) TEXT( GAME_CLIENT_LAUNCH_NAME ), str, NULL, NULL, TRUE, 0, NULL,
```

Comment:
The "os" parts are directly copied from the old steamshim, and I've never written c targeting windows, so i won't  be touching any of it since it seems to work in its current state. feel free to make changes though

Reply:
just realized this also won't account for needing to launch the tvserver. perhaps it's better to switch to a shell script/ bat file as a wrapper instead? i know passing an argument like `-server` was proposed but that feels unintuitive and could introduce more problems...